deepsubtree.go

```diff
@@ -0,0 +1,381 @@
+		fmt.Printf("%s%X:%X (%v)\n", indentPrefix, node.key, node.value, node.height)
+	}
+
+	if node.leftNode != nil {
+		err := dst.printNodeDeepSubtree(node.leftNode, indent+1)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Returns the highest key in the node's subtree
+func (node *Node) getHighestKey() []byte {
+	if node.isLeaf() {
+		return node.key
+	}
+	highestKey := []byte{}
+	if node.rightNode != nil {
+		highestKey = node.rightNode.getHighestKey()
+	}
+	if node.leftNode != nil {
+		leftHighestKey := node.leftNode.getHighestKey()
+		if len(highestKey) == 0 {
+			highestKey = leftHighestKey
+		} else if string(leftHighestKey) > string(highestKey) {
+			highestKey = leftHighestKey
+		}
+	}
+	return highestKey
+}
+
+// Returns the lowest key in the node's subtree
+func (node *Node) getLowestKey() []byte {
+	if node.isLeaf() {
+		return node.key
+	}
+	lowestKey := []byte{}
+	if node.rightNode != nil {
+		lowestKey = node.rightNode.getLowestKey()
+	}
+	if node.leftNode != nil {
+		leftLowestKey := node.leftNode.getLowestKey()
+		if len(lowestKey) == 0 {
+			lowestKey = leftLowestKey
+		} else if string(leftLowestKey) < string(lowestKey) {
+			lowestKey = leftLowestKey
+		}
+	}
+	return lowestKey
+}
+
+func (dst *DeepSubTree) AddProof(proof *ics23.CommitmentProof) error {
+	proof = ics23.Decompress(proof)
+
+	if exist := proof.GetExist(); exist != nil {
+		err := dst.addExistenceProof(exist)
+		if err != nil {
+			return err
+		}
+	} else if nonExist := proof.GetNonexist(); nonExist != nil {
+		err := dst.addExistenceProof(nonExist.Left)
+		if err != nil {
+			return err
+		}
+		err = dst.addExistenceProof(nonExist.Right)
+		if err != nil {
+			return err
+		}
+	}
+
+	return dst.ndb.Commit()
+}
+
+func (dst *DeepSubTree) addExistenceProof(proof *ics23.ExistenceProof) error {
+	leaf, err := fromLeafOp(proof.GetLeaf(), proof.Key, proof.Value)
+	if err != nil {
+		return err
+	}
+	err = dst.ndb.SaveNode(leaf)
+	if err != nil {
+		return err
+	}
+	prevHash := leaf.hash
+	path := proof.GetPath()
+	for i := range path {
+		inner, err := fromInnerOp(path[i], prevHash)
+		if err != nil {
+			return err
+		}
+		prevHash = inner.hash
+
+		has, err := dst.ndb.Has(inner.hash)
+		if err != nil {
+			return err
+		}
+		if !has {
+			err = dst.ndb.SaveNode(inner)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func fromLeafOp(lop *ics23.LeafOp, key, value []byte) (*Node, error) {
+	r := bytes.NewReader(lop.Prefix)
+	height, err := binary.ReadVarint(r)
+	if err != nil {
+		return nil, err
+	}
+	if height != 0 {
+		return nil, errors.New("height should be 0 in the leaf")
+	}
+	size, err := binary.ReadVarint(r)
+	if err != nil {
+		return nil, err
+	}
+	if size != 1 {
+		return nil, errors.New("size should be 1 in the leaf")
+	}
+	version, err := binary.ReadVarint(r)
+	if err != nil {
+		return nil, err
+	}
+	node := &Node{
+		key:     key,
+		value:   value,
+		size:    size,
+		version: version,
+	}
+
+	_, _ = node._hash()
```

Comment:
We should either check the error or have a comment as to why it is ok to ignore the error here.